### PR TITLE
randomSand HW

### PR DESCRIPTION
**Pull request by SheyReiff**
Hello professor Schwimmer, 
My code sometimes work and sometimes I get this error message: 

Execution failed for task ':Main.main()'.
> Process 'command '/Users/sheilareiff/Library/Java/JavaVirtualMachines/openjdk-19.0.2/Contents/Home/bin/java'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
For more on this, please refer to https://docs.gradle.org/8.4/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
BUILD FAILED in 7s
2 actionable tasks: 1 executed, 1 up-to-date

I have been trying to figure it out but I am not sure what the error might be

### CODE REVIEW

**Comment by andrewoid** [comment]
I mentioned there was a bug. Please post what the Exception is when the code crashes.

File: src/main/java/reiff/fallingsand/Main.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            } else if (input.equalsIgnoreCase("q")) {
+                break;
+            }
+        }
+
+        scanner.close();
+    }
+
+    private static void displaySand(Sand sand) {
+        System.out.print("\033[H\033[2J");
```
Comment (andrewoid) [inline src/main/java/reiff/fallingsand/Main.java]: What is this?

File: src/main/java/reiff/fallingsand/Main.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        System.out.flush();
```
Comment (andrewoid) [inline src/main/java/reiff/fallingsand/Main.java]: Is this necessary?

File: src/main/java/reiff/fallingsand/Sand.java [hunk 3]
```diff
@@ -33,16 +46,72 @@ public void put(int x, int y) {
...
+                        field[y][x] = 0;
+                        field[y + 1][x + direction2] = 1;
+                    }
                 }
             }
         }
     }
 
+    public void randomSand(int n) {
+        int freeSpaces = countFreeSpaces();
```
Comment (andrewoid) [inline src/main/java/reiff/fallingsand/Sand.java]: This is called once in the beginning of the program, so it's always going to be large

File: src/main/java/reiff/fallingsand/Main.java (hunk 1 above)
Comment (SheyReiff) [reply src/main/java/reiff/fallingsand/Main.java]: I found it online, it's supposed to clean the console.

File: src/main/java/reiff/fallingsand/Sand.java (hunk 3 above)
Comment (SheyReiff) [reply src/main/java/reiff/fallingsand/Sand.java]: I am not sure I understand what you mean by this.

File: src/main/java/reiff/fallingsand/Main.java (hunk 1 above)
Comment (andrewoid) [reply src/main/java/reiff/fallingsand/Main.java]: Put in a comment on the code what it does and where you found it

File: src/main/java/reiff/fallingsand/Sand.java (hunk 3 above)
Comment (andrewoid) [reply src/main/java/reiff/fallingsand/Sand.java]: There will always be enough free space

**Comment by SheyReiff** [comment]
> I mentioned there was a bug. Please post what the Exception is when the code crashes.

Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 10
	at reiff.fallingsand.Sand.fall(Sand.java:63)
	at reiff.fallingsand.Main.main(Main.java:21)

Is the bug in the code the fact that sometimes the fall method when making the sand fall to either right or left can be out of the field? like accessing a position that does not exist in the field?

File: src/main/java/reiff/fallingsand/Main.java (hunk 1 above)
Comment (SheyReiff) [reply src/main/java/reiff/fallingsand/Main.java]: I realized it is not necessary so I deleted it

**Comment by andrewoid** [comment]
> Is the bug in the code the fact that sometimes the fall method when making the sand fall to either right or left can be out of the field? like accessing a position that does not exist in the field? 

Yes

**Review by andrewoid — APPROVED** [review]

**Comment by andrewoid** [comment]
Please post a screenshot

**Comment by SheyReiff** [comment]
Screenshot 2024-03-13 at 2 14 32 PM

File: src/main/java/reiff/fallingsand/Main.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package reiff.fallingsand;
+
+import java.util.Scanner;
+
+public class Main {
+    public static void main(String[] args) {
+        Scanner scanner = new Scanner(System.in);
+
+        Sand sand = new Sand(10, 50);
```
Comment (andrewoid) [inline src/main/java/reiff/fallingsand/Main.java]: 50, 10

Comment (SheyReiff) [reply src/main/java/reiff/fallingsand/Main.java]: Screenshot 2024-03-13 at 3 04 49 PM

**Pull request merged**